pkg/virt-handler/node-labeller/model.go

```diff
@@ -92,6 +92,9 @@ type Feature struct {
 
 type SEVConfiguration struct {
 	Supported       string `xml:"supported,attr"`
-	Cbitpos         string `xml:"cbitpos"`
-	ReducedPhysBits string `xml:"reducedPhysBits"`
+	CBitPos         uint   `xml:"cbitpos"`
+	ReducedPhysBits uint   `xml:"reducedPhysBits"`
+	MaxGuests       uint   `xml:"maxGuests"`
+	MaxESGuests     uint   `xml:"maxESGuests"`
```

Comment:
If the limits are exceeded, the firmware will just return an error and qemu will fail to start. Ideally we need to consider those limits in the device controller when we setup the plugin for `/dev/sev`. Currently the plugin advertizes the default value (which is the same as for `/dev/kvm` and friends, i.e. `1k`):

https://github.com/kubevirt/kubevirt/blob/3da05ab734d318c7fc3e99fabbb24f5e5de95c4d/pkg/virt-handler/device-manager/device_controller.go#L103-L110

So, answering the initial question: the VM will be scheduled but it will fail to start. I would say that it is smth to fix in a follow-up PR :)

> Does AMD document these limits for specific CPU generations or firmware versions?

I am pretty sure it does. Though I think we are mostly interested in getting those during runtime.

> Does the entire stack keep the counts up to date as VMs are scheduled on a node? What would it look like for Kubevirt to query this each time we try to launch an SEV/SEV-ES guest? If the count isn't reliable, can kubevirt keep internal counts of the number of SEV/SEV-ES guests?

I think there is no need for KubeVirt to query it each time (and I dont think there is an API for getting the current ammount of SEV guests). As mentioned above I would rather offload the accounting to the kuberbetes device mgmt.

Reply:
Thanks for the clarifications!

Indeed this sounds like something to tackle in a follow-up PR.

At a first glance it makes sense to me that virt-handler would have an internal counter and would mark the node with a label if no more SEV workloads can be scheduled.